File: MC_Graph simulation/MC_variable_distributions.m
function [Fd]=MC_variable_distributions(n,varargin)
% function that simulates a scale-free network with preferential attachment
% with fitness preferences as in WP_LEM.
%INPUT:
%         n - total number of nodes (size of the financial bank system)
%         mu - the power law coeff. for the fitness distribution (bank assets)
%         ad = average degree for directed networks ad=L/N
%         alfabeta=[alpha beta] - the power law parameters of the linking function
%         xmin - the min size of the size distribution (assets size)
%         xmax - the max size of the size distribution (assets size) (in truncated pareto)
%  varargin -> decide the distribution of the fitness variables (assets size of banks)
%
%OUTPUT
%        Fd = the size distribution from the fitness distribution
% Example:
%        fitness_density_MFa(n,'powerlaw',mu);
%        fitness_density_MFa(n,'xmin',xmin,'powerlaw',mu);
%        fitness_density_MFa(n,'xmin',xmin,'truncated',mu,xmax);
%        fitness_density_MFa(n,'xmin',xmin,'powercutoff',mu,lambda);
%        fitness_density_MFa(n,'exponential',lambda);
%        fitness_density_MFa(n,'stretched',lambda,bs);
% FabioVanni©2015



%default values 
xmin=5;
xmax=50;
mu=2;

%% ARGUMENTS selections
% parse command-line parameters; trap for bad input
if isempty(varargin), disp(' !!! ATTENTION !!! Using as power law fitness with mu=2'),
    type='PL';
    xmin=1;mu=2;
end
i=1; 
while i<length(varargin), 
  argom = 1; 
  if ischar(varargin{i}), 
    switch varargin{i},
        case 'xmin',            xmin = varargin{i+1}; i = i + 2;
        case 'powerlaw',     type = 'PL'; mu  = varargin{i+1}; i = i + 1;
        case 'truncated',    type = 'TC'; mu = varargin{i+1}; xmax = varargin{i+2};i = i + 2;
        case 'powercutoff',          type = 'PC'; mu  = varargin{i+1}; lambda = varargin{i+2}; i = i + 2;
        case 'exponential',     type = 'EX'; lambda = varargin{i+1}; i = i + 3;
        case 'stretched',       type = 'ST'; lambda = varargin{i+1}; bs = varargin{i+2}; i = i + 4;
        otherwise, argom=0;disp(['(scalefree) !!! INVALID INPUTS !!! , using default truncated pareto' ]);i=5 ; 
            n=250;xmin=5;xmax=50;mu=2; type='TC';
    end
  end
end

switch type
    case 'EX',% EXPONENTIAL (random graph)
        x = xmin - (1/lambda)*log(1-rand(n,1));
    case 'TC', % TRUNCATED Pareto
            alfa=mu-1;
            u=rand(n,1);
            l=xmin;
            h=xmax;
            if l<h,
            x=(-(u*h^alfa - u*l^alfa - h^alfa)./( h^alfa * l^alfa ) ).^(-1/alfa);
            else error('wrong ranges')
            end
    case 'ST', % stretched exponential truncation
        x = (xmin^bs - (1/lambda)*log(1-rand(n,1))).^(1/bs);
    case 'PC', % Exponential cutoff
        x = [];
        y = xmin - (1/lambda)*log(1-rand(10*n,1));
            while true
                y(rand(10*n,1)>=(y./xmin).^(-mu)) = [];
                x = [x; y];
                q = length(x)-n;
                if (q==0), break;
                end;
                if (q>0),
                    r = randperm(length(x));
                    x(r(1:q)) = [];
                    break;
                end;
                if (q<0),
                    y = xmin - (1/lambda)*log(1-rand(10*n,1));
                end;
            end; 
    case 'PL', % POWERLAW with xmin
         x = xmin*(1-rand(n,1)).^(-1/(mu-1));
    otherwise, x = xmin*(1-rand(n,1)).^(-1/(mu-1));
end;
%%
  
Fd=x; 

File: MC_Graph simulation/MC_variable_distributions.py
import numpy as np

def MC_variable_distributions(n, *args):
    # Default values
    xmin = 5
    xmax = 50
    mu = 2
    
    # ARGUMENTS selections
    type = 'PL'
    
    i = 0
    while i < len(args):
        argom = 1
        if isinstance(args[i], str):
            if args[i] == 'xmin':
                xmin = args[i + 1]
                i += 2
            elif args[i] == 'powerlaw':
                type = 'PL'
                mu = args[i + 1]
                i += 2
            elif args[i] == 'truncated':
                type = 'TC'
                mu = args[i + 1]
                xmax = args[i + 2]
                i += 3
            elif args[i] == 'powercutoff':
                type = 'PC'
                mu = args[i + 1]
                lambda_ = args[i + 2]
                i += 3
            elif args[i] == 'exponential':
                type = 'EX'
                lambda_ = args[i + 1]
                i += 2
            elif args[i] == 'stretched':
                type = 'ST'
                lambda_ = args[i + 1]
                bs = args[i + 2]
                i += 3
            else:
                argom = 0
                print('(scalefree) !!! INVALID INPUTS !!! , using default truncated pareto')
                i = 5
                n = 250
                xmin = 5
                xmax = 50
                mu = 2
                type = 'TC'
        
        if argom == 0:
            break
    
    if type == 'EX':
        # EXPONENTIAL (random graph)
        x = xmin - (1 / lambda_) * np.log(1 - np.random.rand(n, 1))
    elif type == 'TC':
        # TRUNCATED Pareto
        alfa = mu - 1
        u = np.random.rand(n, 1)
        l = xmin
        h = xmax
        if l < h:
            x = (-((u * h ** alfa - u * l ** alfa - h ** alfa) / (h ** alfa * l ** alfa)) ) ** (-1 / alfa)
        else:
            raise ValueError('Wrong ranges')
    elif type == 'ST':
        # stretched exponential truncation
        x = (xmin ** bs - (1 / lambda_) * np.log(1 - np.random.rand(n, 1))) ** (1 / bs)
    elif type == 'PC':
        # Exponential cutoff
        x = np.array([])
        y = xmin - (1 / lambda_) * np.log(1 - np.random.rand(10 * n, 1))
        while True:
            y = y[y >= (y / xmin) ** (-mu)]
            x = np.concatenate((x, y), axis=0)
            q = len(x) - n
            if q == 0:
                break
            if q > 0:
                r = np.random.permutation(len(x))
                x = np.delete(x, r[:q])
                break
            if q < 0:
                y = xmin - (1 / lambda_) * np.log(1 - np.random.rand(10 * n, 1))
    elif type == 'PL':
        # POWERLAW with xmin
        x = xmin * (1 - np.random.rand(n, 1)) ** (-1 / (mu - 1))
    else:
        x = xmin * (1 - np.random.rand(n, 1)) ** (-1 / (mu - 1))
    
    Fd = x
    
    return Fd

